src/main/scala/is/hail/rvd/OrderedRVD.scala

```diff
@@ -566,15 +622,57 @@ class OrderedRVD(
     partitioner,
     this.crdd.czip(that.crdd, preservesPartitioning = true)(zipper))
 
+  // New key type must be prefix of left key type. 'joinKey' must be prefix of
+  // both left key and right key. 'zipper' must take all output key values from
+  // left iterator, and be monotonic on left iterator (it can drop or duplicate
+  // elements of left iterator, or insert new elements in order, but cannot
+  // rearrange them), and output region values must conform to 'newTyp'. The
+  // partitioner of the resulting OrderedRVD will be left partitioner truncated
+  // to new key. Each partition will be computed by 'zipper', with corresponding
+  // partition of 'this' as first iterator, and with all rows of 'that' whose
+  // 'joinKey' might match something in partition as the second iterator.
+  def alignAndZipPartitions(
+    newTyp: OrderedRVDType,
+    that: OrderedRVD,
+    joinKey: TStruct
+  )(zipper: (RVDContext, Iterator[RegionValue], Iterator[RegionValue]) => Iterator[RegionValue]
+  ): OrderedRVD = {
+    require(newTyp.kType isPrefixOf this.typ.kType)
+    require(joinKey isPrefixOf this.typ.kType)
+    require(joinKey isPrefixOf that.typ.kType)
+
+    val left = this.truncateKey(newTyp.key)
+    val coarsenedPartitioner = this.partitioner.coarsen(joinKey.size)
+    OrderedRVD(
+      typ = newTyp,
+      partitioner = left.partitioner,
+      crdd = left.crddBoundary.czipPartitions(
+        that.constrainToOrderedPartitioner(
+          that.typ.copy(
+            partitionKey = that.typ.key.take(coarsenedPartitioner.partitionKey.getOrElse(that.typ.key.length)),
+            key = that.typ.key.take(coarsenedPartitioner.kType.size)),
+          coarsenedPartitioner
+        ).crddBoundary
```

Comment:
Should I expect the partitions of `this.truncateKey(newTyp.key)` to match those of `this.partitioner.coarsen(joinKey.size)` because neither actually change the range bounds?

Also, it looks like `constrainToOrderedPartitioner` basically duplicates any source partitions that now lie in distinct partitions, is that right?

Reply:
I think I may have to answer this question in person...